docs/writing-tests.md

```diff
@@ -0,0 +1,60 @@
+  - [Bad example](https://github.com/alphagov/smokey/blob/aaa0522666d49db116b939276f0e0969357e3ed4/features/collections.feature#L29) - checks an arbitrary, non-critical page ([lower level test exists](https://github.com/alphagov/collections/blob/cc94525a05b2651fee621b588e530aca8e1f77e4/features/viewing_browse.feature))
+  - [Bad example](https://github.com/alphagov/smokey/blob/aaa0522666d49db116b939276f0e0969357e3ed4/features/finder_frontend.feature#L45) - missing a lower level test in the app itself ([example test to copy](https://github.com/alphagov/finder-frontend/blob/d70a01903e2813719e7d5adbf79fbcb1d49335fb/spec/controllers/email_alert_subscriptions_controller_spec.rb#L363))
+
+Do not duplicate tests. For example, ["Email signup from an organisation home page"](https://github.com/alphagov/smokey/blob/aaa0522666d49db116b939276f0e0969357e3ed4/features/collections.feature#L47) duplicates ["Email signup from a taxon page"](https://github.com/alphagov/smokey/blob/aaa0522666d49db116b939276f0e0969357e3ed4/features/collections.feature#L53): both cover data transfer with Email Alert API.
+
+Do not add tests that can be implemented in other ways. We want this to be the tip of the [testing pyramid](https://martinfowler.com/bliki/TestPyramid.html), not a [testing ice cream cone](http://saeedgatson.com/the-software-testing-ice-cream-cone/). Instead of adding a test here, you can:
+
+- Add or extend [health checks for application infrastructure](https://github.com/alphagov/govuk_app_config/blob/main/docs/healthchecks.md). For example, drafting a document in a publishing app involves connecting to infrastructure like a database or AWS S3.
+
+- Add [contract tests](https://docs.publishing.service.gov.uk/manual/pact-broker.html) to cover a chain of APIs. For example, subscribing to an email involves communication with Email Alert API and Content Store, which can be tested via GDS API Adapters.
+
+- Add unit or integration tests to cover in-app behaviour. For example, the way each document format is rendered can be tested in the frontend app that does the rendering.
+
+In contrast, the tests in this repo provide coverage at a higher level. We should be willing to page someone when tests are failing, once they are reliable enough.
+
+## Adding new tests
+
+We use feature files to describe single applications (e.g. [`whitehall`](https://github.com/alphagov/whitehall), [`frontend`](https://github.com/alphagov/frontend)), as well as [cross-application behaviour](https://github.com/alphagov/smokey/blob/main/features/gov_uk.feature).
+
+Try to use existing step definitions e.g. [`When I visit "/some/path"`](https://github.com/alphagov/smokey/blob/b656e65ef6853d65de3e24a7e884911054c8bf7d/features/step_definitions/smokey_steps.rb#L70-L72). If a more specific step is required, name the `_steps.rb` file after the feature file i.e. `myfeature_steps.rb`.
+
```

Comment:
We have a lot of tests that seem to be concerned with the implementation of GOV.UK, rather than just testing the user functionality (e.g. Check homepage is served by Varnish). Once we start testing configuration our test become more fragile and they aren't testing what we actually care about (i.e. that we can just load the homepage). For example, if we change the configuration but the feature still works when end up breaking tests.

Would be good if we had guidance that new tests are written to test user features, rather than configuration.

Reply:
Hmm...this is tricky one.

I don't think we should change the guidance as it stands until an alternative exists. I do agree that the mixture of user-focussed and non-user-focussed tests makes the overall approach to writing tests a bit confusing. One of the changes I plan to make after this PR is to reorganise some of the tests, so the "config" ones should be a bit more contained.

Having audited them all, I don't think the existing "config" tests are fragile e.g.

- "Check homepage is served by Varnish (CDN)" is unlikely to be affected by changes to the home page.
- "Check redirect from bare domain to www.gov.uk is working for HTTP" should also be robust longterm.

I think I get your point though: by writing tests that assert at a lower level, we set a precedent for tests that _are_ fragile, because they're too granular. It's worth noting this can happen just as much with (so called) user-focussed tests ([example](https://github.com/alphagov/smokey/pull/965/files#diff-d806eef09bf240d54b61cc2de907d6cb57806d48329fc29884f0c0cc5556d7f6R58)). In other words, I don't think the "config" tests represent a real problem in the way you describe.

However, in the longterm we should have a separate suite of tests for things like the CDN, where we fully test everything we care about - "vital" or not. But if we disallow new "config" tests now we effectively force a developer to create the alternative, which might happen, or the test might not get written if they think it's too difficult.

I'm seeing this PR as a step - a prerequisite, really - towards these kinds of changes.

What do you think? Did I understand you correctly?

Reply:
Ahh right, maybe the name for that example I gave was confusing - it actually tests whether the pages is cached (which is fine), rather than checking that its being served explicitly by Varnish. That would have created a fragile test - as we could change the caching layer to another service other than varnish, and it would have broken the test, but we'd still be caching the request (the thing we care about).

I can't find another test that tests configuration like this with a quick glance over. So don't want this comment to block the PR.